### PR TITLE
Fix potential divide by zero

### LINKED ISSUE #50761: rustc panic with divide by zero
thread 'rustc' panicked at 'attempt to divide by zero', librustc_trans/abi.rs:430:26

Stripped down version of code to reproduce
```rust
mod a {
    pub trait A {}
}

mod b {
    pub struct Builder {}

    pub fn new() -> Builder {
        Builder {}
    }

    impl Builder {
        pub fn with_a(&mut self, _a: fn() -> ::a::A) {}
    }
}

pub use self::b::new;
```

## Meta `rustc --version --verbose`:
```
rustc 1.26.0 (a77568041 2018-05-07)
binary: rustc
commit-hash: a7756804103447ea4e68a71ccf071e7ad8f7a03e
commit-date: 2018-05-07
host: x86_64-unknown-linux-gnu
release: 1.26.0
LLVM version: 6.0
```
stack backtrace:
```
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:207
   3: std::panicking::default_hook
             at libstd/panicking.rs:223
   4: core::ops::function::Fn::call
   5: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:403
   6: std::panicking::begin_panic_fmt
             at libstd/panicking.rs:349
   7: rust_begin_unwind
             at libstd/panicking.rs:325
   8: core::panicking::panic_fmt
             at libcore/panicking.rs:72
   9: core::panicking::panic
             at libcore/panicking.rs:51
  10: rustc_trans::abi::CastTarget::llvm_type
  11: rustc_trans::abi::FnType::llvm_type
  12: <rustc::ty::layout::TyLayout<'tcx> as rustc_trans::type_of::LayoutLlvmExt<'tcx>>::llvm_type
  13: rustc_trans::abi::FnType::llvm_type
  14: rustc_trans::declare::declare_fn
  15: rustc_trans::trans_item::predefine_fn
  16: rustc_trans::trans_item::MonoItemExt::predefine
  17: rustc_trans::base::compile_codegen_unit
  18: rustc::dep_graph::graph::DepGraph::with_task_impl
  19: rustc_errors::Handler::track_diagnostics
  20: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::cycle_check
  21: rustc::ty::maps::<impl rustc::ty::maps::queries::compile_codegen_unit<'tcx>>::force
  22: rustc::ty::maps::<impl rustc::ty::maps::queries::compile_codegen_unit<'tcx>>::try_get
  23: rustc::ty::maps::TyCtxtAt::compile_codegen_unit
  24: rustc::ty::maps::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::compile_codegen_unit
  25: rustc_trans::base::trans_crate
  26: <rustc_trans::LlvmTransCrate as rustc_trans_utils::trans_crate::TransCrate>::trans_crate
  27: rustc::util::common::time
  28: rustc_driver::driver::phase_4_translate_to_llvm
  29: rustc_driver::driver::compile_input::{{closure}}
  30: <std::thread::local::LocalKey<T>>::with
  31: <std::thread::local::LocalKey<T>>::with
  32: rustc::ty::context::TyCtxt::create_and_enter
  33: rustc_driver::driver::compile_input
  34: rustc_driver::run_compiler_impl
  35: syntax::with_globals
```

### PR DESCRIPTION
**Pull request by cjkenn**
This should fix #50761

I had trouble reproducing with the provided code, but looking at the stack trace would indicate that this code is the likely cause. I made a number of assumptions here, because I don't have enough context on how the register size is set:

1. I assumed `rest.unit.size.bytes()` can be 0, and it's ok if it's set to 0 before this function is called
2. I assumed that if `rest.unit.size.bytes()` is 0, that we want `rest_count` to also be 0.

### CODE REVIEW

File: src/librustc_trans/abi.rs [hunk 1]
```diff
@@ -127,7 +127,12 @@ impl LlvmType for Reg {
...
     fn llvm_type(&self, cx: &CodegenCx) -> Type {
         let rest_ll_unit = self.rest.unit.llvm_type(cx);
-        let rest_count = self.rest.total.bytes() / self.rest.unit.size.bytes();
+        let rest_count = if self.rest.unit.size.bytes() == 0 {
+            0
+        } else {
+            self.rest.total.bytes() / self.rest.unit.size.bytes()
+        };
+
         let rem_bytes = self.rest.total.bytes() % self.rest.unit.size.bytes();
```
Comment (kennytm) [inline src/librustc_trans/abi.rs]: If `y == 0`, then `x % y` will also panic, please move this `rem_bytes` calculation inside the conditional as well.

**Review by kennytm — DISMISSED** [review]
Thanks for the fix!

Could you add the test case from #50761 so we could ensure the bug is indeed fixed?

**Comment by cjkenn** [comment]
Should the test file be added to codegen tests? I've added a ui test because we just want to ensure that the code from the issue passes compilation right?

**Comment by cjkenn** [comment]
Sorry for the commit spam after rebasing :( I can remove or squash them when I get a sec again

**Comment by michaelwoerister** [comment]
@cjkenn Yes, please `:)`

**Comment by cjkenn** [comment]
Err hopefully this is better after some more git fiddling -_-... I didn't squash them but the diffs are right.

**Comment by kennytm** [comment]
@cjkenn Could you do a rebase instead of merge? 

```console
$ git checkout master
$ git pull origin master --ff-only    # synchronize master branch with this repository
$ git checkout div-by-zero
$ git rebase master      # rebase on top of the latest master
$ git rebase -i master   # do more cleanup e.g. squash everything down to a single commit
$ git push cjkenn div-by-zero --force-with-lease
```

**Comment by cjkenn** [comment]
Ok, I think I've FINALLY got it... Thanks for the git help!

File: src/test/ui/issue-50761.rs [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Confirm that we don't accidently divide or mod by zero in llvm_type
+
+// compile-pass
+
+#![feature(test)]
```
Comment (kennytm) [inline src/test/ui/issue-50761.rs]: This line should be removed?

**Review by kennytm — APPROVED** [review]
r=me after a minor nit is fixed.

**Pull request merged**